Logstash Getting started with Kubernetes Guide (#14655)
elastic/logstash: docsk8s/quick-start/ls-k8s-quick-start.asciidoc

Before you start, there are a few things you'll need:

> . A running Kubernetes cluster - For local/single node testing we recommend using link:https://minikube.sigs.k8s.io[Minikube], which allows you to easily run a single node Kubernetes cluster on your system. Check the `Get Started!` section for install and set up instructions.

Author: Removed stray backtick from your suggestion
```suggestion
* *A running {k8s} cluster.*  For local/single node testing we recommend using https://minikube.sigs.k8s.io[Minikube], which allows you to easily run a single node {k8s} cluster on your system. 
Check the minikube https://minikube.sigs.k8s.io/docs/start/[Get Started!] section for install and set up instructions.
```

subject=/C=EU/ST=NA/O=Elastic/CN=ClientName
Getting CA Private Key
--

Your `logstash-k8s-qs/cert` folder should now contain a set of certificate files, including `client` certificates for Filebeat and Metricbeat, and `server` certificates for Logstash.

The parent `logstash-k8s-qs` directory also has a new `001-secret.yaml` resources file that stores a hash of the client and server certificates.

image::./images/gs-cert-files.png[generated CA certificate files]

====

[float]
[[qs-create-kubernetes-cluster]]
=== Create Kubernetes Cluster

As part of this configuration, we will be setting up Elastic Stack components and Logstash

[float]
[[qs-create-elastic-stack]]
==== Create Elastic Stack components

Now that your environment and certificates are set up, it's time to create an Elastic Stack. We will be creating:

* Elasticsearch - you know, for search
* Kibana - for data visualization
* Filebeat - to monitor container logs
* Metricbeat - to monitor logstash and send stack monitoring data.

> as well as the secret definition containing the keys and certificates we generated earlier.

Author: Yes, this should be another bullet I think

demo-es-default-0                       1/1     Running   0              41s
demo-kb-d7f585494-vbf6s                 1/1     Running   0              39s
elastic-operator-0                      1/1     Running   4 (164m ago)   13d
--

[float]
[[qs-set-up-logstash]]
==== Set up Logstash

Now we have an elastic stack to communicate with, let's set up Logstash.

When setting up Logstash in Kubernetes, we will typically use <<qs-configmap, ConfigMaps>> to set up Logstash configurations, and our pipeline definitions - see <<ls-k8s-configuration-files, Logstash Configuration files in Kubernetes>> for more details on how to configure Logstash in Kubernetes.

> Then, we'll create the <<qs-deployment, deployment definition>> for {ls}, including memory, CPU resources, the container ports, timeout settings, and similar, and the <<qs-service, Service definition>>, opening up ports on the logstash pods to the internal metricbeat (for stack monitoring) and filebeat in this instance

Author: How about something like:

```suggestion
Next we'll create the <<qs-deployment, deployment definition>> for {ls}. 
It includes settings for memory, CPU resources, which ports the container uses, timeouts, etc.
It also includes the <<qs-service, Service definition>> that opens ports to the internal {metricbeat} instance (for stack monitoring) and a {filebeat} instance for this example. ```

> Then, we'll create the <<qs-deployment, deployment definition>> for {ls}, including memory, CPU resources, the container ports, timeout settings, and similar, and the <<qs-service, Service definition>>, opening up ports on the logstash pods to the internal metricbeat (for stack monitoring) and filebeat in this instance

Reviewer: Also, the three backticks at the end of the suggested text look like they might render if you commit this suggestion.  Please check it.

> Then, we'll create the <<qs-deployment, deployment definition>> for {ls}, including memory, CPU resources, the container ports, timeout settings, and similar, and the <<qs-service, Service definition>>, opening up ports on the logstash pods to the internal metricbeat (for stack monitoring) and filebeat in this instance

Author: Checked, and fixed

> Then, we'll create the <<qs-deployment, deployment definition>> for {ls}, including memory, CPU resources, the container ports, timeout settings, and similar, and the <<qs-service, Service definition>>, opening up ports on the logstash pods to the internal metricbeat (for stack monitoring) and filebeat in this instance

Author: I like "and so forth"

 
[source,sh]
--
kubectl delete service,pods,deployment,configmap,secret,beat,elasticsearch,kibana -l app=logstash-demo
--

[float]
[[qs-next-steps]]
=== Next Steps

[float]
[[qs-external-elasticsearch]]
==== Send logs to an External Elasticsearch

Rather than sending to an elasticsearch cluster in the same Kubernetes cluster as logstash, you may want to send data to an Elastic cloud instance.

[float]
[[qs-send-to-elastic-cloud]]
===== Sending to Elastic Cloud

To connect to Elastic cloud, we will only need the logstash based components - no need to include the `elasticsearch` or `kibana` components from the walkthrough

We will need to amend the `Deployment`/`StatefulSet` to set `CLOUD_ID` and `API_KEY` environment variables with the appropriate value for your cloud instance.

One way to do this is to create a link:https://kubernetes.io/docs/concepts/configuration/secret/[secret] to store `CLOUD_ID` and `API_KEY`:

[source,yaml]
--
apiVersion: v1
kind: Secret
metadata:
  name: ess_secret
type: Opaque
data:
  cloud_id: PENMT1VEX0lEPg== <1>
  password: PEFQSV9LRVk+
--

> <1> base64 representation of `cloud_id` and `api_key` for your elastic cloud instance - created by

Reviewer: ```suggestion
<1> base64 representation of `cloud_id` and `api_key` for your elastic cloud instance - created using:
```

+
[source,sh]
--
echo -n '<CLOUD_ID>' | base64
echo -n '<API_KEY>' | base64
--

Mount the secrets in the `Deployment`/`StatefulSet`

[source,yaml]
--
env:
  - name: CLOUD_ID
      valueFrom:
        secretKeyRef:
          name: ess_secret
          key: cloud_id
    - name: API_KEY
      valueFrom:
        secretKeyRef:
          name: ess_secret
          key: api_key

--

> And we will amend the pipeline definition `ConfigMap` to change the destination of the elasticsearch output to the cloud instance.

Reviewer: ```suggestion
Let's amend the pipeline definition `ConfigMap` to change the destination of the {es} output to the cloud instance.
```

[source,yaml]
--
    output {
      elasticsearch {
        cloud_id => "CLOUD_ID"
        api_key => "API_KEY"
        ssl => true
      }
--

[float]
[[qs-scale-logstash]]
==== Scale Logstash with Horizontal Pod Autoscaler

For a simple Logstash setup, without <<ls-k8s-persistent-storage, persistent storage>>, or <<ls-k8s-design-for-plugins, plugins that require the storing of local state>>, we can introduce a simple <<qs-autoscaler, horizontal pod autoscaler>>.

> Apply the autoscaler as follows:

Reviewer: ```suggestion
Apply the autoscaler:
```

[source,bash]
--
kubectl apply -f "004-hpa.yaml"
--

> NOTE: When using more than one logstash pod, use the link:https://www.elastic.co/guide/en/beats/metricbeat/current/configuration-autodiscover.html#_kubernetes[autodiscover] features of beats to monitor the different logstash pods, otherwise only one logstash pod will be monitored.

Author: Do you think it makes sense to link to https://github.com/elastic/logstash/blob/05bfaff799d3cb52b2d0af51d9a9a51f15a0501f/docsk8s/administering/ls-k8s-stack-monitoring.asciidoc#L17 where Kaise mentioned autodiscover in the context of stack monitoring?

> NOTE: When using more than one logstash pod, use the link:https://www.elastic.co/guide/en/beats/metricbeat/current/configuration-autodiscover.html#_kubernetes[autodiscover] features of beats to monitor the different logstash pods, otherwise only one logstash pod will be monitored.

Reviewer: Sure

> NOTE: When using more than one logstash pod, use the link:https://www.elastic.co/guide/en/beats/metricbeat/current/configuration-autodiscover.html#_kubernetes[autodiscover] features of beats to monitor the different logstash pods, otherwise only one logstash pod will be monitored.

Author: ```suggestion
NOTE: If you are using more than one {ls} pod, use the https://www.elastic.co/guide/en/beats/metricbeat/current/configuration-autodiscover.html#_kubernetes[beats autodiscover] features to monitor them. Otherwise, only one {ls} pod is monitored.
See the <<stack monitoring with ECK,monitor-with-ECK>> docs for details on how to use autodiscover with {metricbeat} and {ls}.

```
